test/base/testing_brave_browser_process.cc

```diff
@@ -0,0 +1,63 @@
+/* Copyright (c) 2021 The Brave Software Team. Distributed under the MPL2
+ * license. This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <utility>
+
+#include "brave/components/brave_shields/browser/ad_block_service.h"
+#include "brave/components/ipfs/buildflags/buildflags.h"
+#include "brave/components/tor/buildflags/buildflags.h"
+#include "brave/test/base/testing_brave_browser_process.h"
+
+namespace tor {
+class BraveTorClientUpdater;
+}
+
+namespace ipfs {
+class BraveIpfsClientUpdater;
+}
+
+// static
+TestingBraveBrowserProcess* TestingBraveBrowserProcess::GetGlobal() {
+  return static_cast<TestingBraveBrowserProcess*>(g_brave_browser_process);
+}
+
+// static
+void TestingBraveBrowserProcess::CreateInstance() {
+  DCHECK(!g_brave_browser_process);
+  TestingBraveBrowserProcess* process = new TestingBraveBrowserProcess;
+  g_brave_browser_process = process;
+}
+
+// static
+void TestingBraveBrowserProcess::DeleteInstance() {
+  g_brave_browser_process = nullptr;
```

Comment:
Looks like you're correct - but in the process of fixing this I realized it's not quite so simple as `delete`ing the instance.

The destructor for `AdBlockBaseService` requires access to the `BraveComponent::Delegate` because it calls `GetTaskRunner()`. However, this will not work in unit tests because the delegate is only constructed within the scope of each test, so it is destroyed before the adblock service.

https://github.com/brave/brave-core/blob/552554306ae88fe9f776b3b3add00788a0f39b8f/components/brave_shields/browser/ad_block_base_service.cc#L110-L112

There's a few ways I can think of to solve this, but none of them stand out as obviously correct to me:
- expose a method on `TestingBraveBrowserProcess` to reclaim the adblock service so that it can be deleted manually
- store the delegate in a field of `TestingBraveBrowserProcess`
- Remove the task runner access from the adblock service destructor (i.e. just delete it instantly? is there any reason it needs to be there?)

Reply:
Fixed by calling `TestingBraveBrowserProcess::DeleteInstance()` in the test suite's `TearDown()`. It should normally handled automatically by the subclassed `DeleteInstance` method, but there are a few unit tests in Chromium that must call it manually like I've done here.